ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/plugin/ExecutionEnvironmentSection.java

```diff
@@ -262,17 +274,53 @@ private void handleRemove() {
 	}
 
 	private void handleAdd() {
-		ElementListSelectionDialog dialog = new ElementListSelectionDialog(PDEPlugin.getActiveWorkbenchShell(), new EELabelProvider());
+		ElementListSelectionDialog dialog = showDialog(true);
+		if (dialog.open() == Window.OK) {
+			addExecutionEnvironments(dialog.getResult());
+		}
+	}
+
+	private void handleUpgrade() {
+		RequiredExecutionEnvironmentHeader header = getHeader();
+		IExecutionEnvironment[] executionEnvironments = JavaRuntime.getExecutionEnvironmentsManager()
+				.getExecutionEnvironments();
+		if (executionEnvironments.length > 0) {
+			ElementListSelectionDialog dialog = showDialog(false);
+			if (dialog.open() == Window.OK) {
+
+				List<String> addedExecutionEnvironments = header.getEnvironments();
+				// Remove existing EE ( Upgrade is allowed only when single Java
+				// EE is present in the header)
+				getHeader().removeExecutionEnvironment(addedExecutionEnvironments.get(0));
```

Comment:
The variable name could be a bit more precise and the comment can be updated that all single EEs are upgradable.
```suggestion
			List<String> existingEEs = header.getEnvironments();
			// Remove existing EE (Upgrade is allowed only if a single EE is
			// present in the header)
			getHeader().removeExecutionEnvironment(existingEEs.get(0));
```